### ISSUE TITLE
[AMA] Auto Completion

### ISSUE THREAD

**Issue by kingToolbox** [kingToolbox/WindTerm#773]
**I'm happy to answer any questions and really look forward to getting any feedback and ideas from you!**

[Auto completion](https://kingtoolbox.github.io/2022/05/18/auto_completion/) is one of the most important features of `WindTerm`. I plan to complete the entire feature development in three development cycles of `v2.5`, `v2.6` and `v2.7`. `Auto completion` includes but are not limited to:

- Auto-complete `command name`, `command options` and `command parameters`.
- Auto-complete `command histories`.
- Auto-complete user-defined `command snippets`.
- Better auto-completion based on context and history.
- Better auto-completion based on `file system` and `system environment`.

I have written nearly **4,000** command configuration files, but this is obviously **not enough**. After the design is finalized, I look forward to more people participating in the writing work. All command configurations, including the **4000** commands I have written, will be released under the [Creative Commons Attribution 4.0 International Public License](https://creativecommons.org/licenses/by/4.0/legalcode). Not only `WindTerm` can use these commands, any terminal can use them too.

The [WindTerm_2.5.0_Prerelease_1](https://github.com/kingToolbox/WindTerm/releases/tag/2.5-prerelease) just released contains some auto-completion features. Welcome to experience it. Thank you.

**Comment by samulo911** [comment]
1.Need an option to turn off the auto completion function since it now doesn't support linux alias.For example, I have an alias "ll" for "ls -lh" and it matches "kill" command with  auto completion.I have to press extra ESC key to close the auto completion frame.
2.If you type "ls" as quickly as you can,even though it full-text matches command "ls",auto completion function sometimes will turn it into "less" which is the first candidate of single letter "l".

**Comment by luantu** [comment]
how to enable this feature ?

**Comment by kingToolbox** [comment]
Just download [WindTerm_2.5.0_Prerelease_1](https://github.com/kingToolbox/WindTerm/releases/tag/2.5-prerelease) and type :smile: you can watch this [intro video](https://kingtoolbox.github.io/2022/05/18/auto_completion/) to learn more.

**Comment by luantu** [comment]
> 

I just replace with the version 2.5.0 .but nothing happened  
running on windows 10, and ssh to my ubuntu 18.04

https://user-images.githubusercontent.com/5064819/169219465-b0b18f02-f7d2-4164-bd98-bf45faf136f9.mp4

**Comment by luantu** [comment]
I just found that it works in another telnet session , but did not work in ssh session mentioned above.  >_<

![image](https://user-images.githubusercontent.com/5064819/169227978-af3d3923-1496-4b6e-b7eb-928953e5a15b.png)

**Comment by eugenov** [comment]
@kingToolbox and how to disable it?

**Comment by kingToolbox** [comment]
@luantu 
Since the text in `WindTerm` does not have any rules and grammatical constraints, it is indeed very difficult to process the text. The current design relies heavily on the command prompt. So once you use tools such as `zsh`, you will not be able to use the `auto completion` because `WindTerm` cannot find the command prompt. And there is no way to find a good solution for this. :disappointed:

**Comment by kingToolbox** [comment]
@eugenov 
I'm usually used to testing new features in `Prerelease` versions. So there is currently no option to disable it. But I will provide such an option in the official version. I will also try my best to make `auto completion` perfect enough so that you won't think about closing it. :smile:

**Comment by shengkui** [comment]
It's boring sometimes, waiting a disable option.

**Comment by kingToolbox** [comment]
Yes, I've thought about it after @eugenov asked this question. In the next prerelease version, `auto completion` will only be available for configured commands. This can reduce unnecessary interference very well. After all, there are not enough command configurations that have been completed.

Tips:
If you enter a command, such as `vi`, and there is no matching command in the autocomplete list. You don't need to press the `esc` key to cancel the autocomplete list, you can directly continue to enter a `space` to automatically close the list.

**Comment by eugenov** [comment]
Typing "ll"-"enter" in directory which files you want to list ends up with "kill". Usually it is one of very quick-typing commands, so I have no time to think about it

**Comment by kingToolbox** [comment]
@eugenov 
Creating an empty file `ll.command` or `ls,ll.command` in the `WindTerm_install_path/terminal/commands/linux` can temporarily circumvent your problem. 

This is the first prerelease version after all, so there will indeed be a lot of shortcomings. But in the next prerelease version, I will add some required settings, as well as a better `auto completion`, including better support for commands without parameters, in order to meet the needs of daily use.

**An unhelpful autocomplete is not a good autocomplete I want.** :smile:

**Comment by eugenov** [comment]
Oh, I see that I can write autocomplete commands by myself? Can I create user commands and lexers in profile directory?
It is a great news! Looking forward updates and features, thanks for your hard work!

**Comment by kingToolbox** [comment]
Yes, commands are fully customizable. There are so many commands, if I need to write configuration for each, I don't think I have time to eat and sleep. :sweat_smile: 

Unfortunately, like other features, such as `keymaps`, `themes`, etc., it is not yet possible to support users to write their own configuration files in the `./wind/profiles` folder. This is because many designs are still in flux. If I change the design and then a lot of users need to rewrite or modify their config files, That would be a disaster for me. I expect customization will be available starting from the `2.8` version. And I will provide all customizations before the `3.0` version, and fix all known bugs in `issues` as much as possible.

Starting from version `3.0`, a plugin system and a plugin download platform will be developed, which will bring a better future to `WindTerm`.

**Comment by eugenov** [comment]
> There are so many commands, if I need to write configuration for each, I don't think I have time to eat and sleep.

imao users need only basic commands where autocomplete realy helps, the rest we could write by ourselves. Just give us a manual and a easy way to customize them. So let's save a lot of time for important things :)

**Comment by iciness** [comment]
How to disable auto completion?

**Comment by zdltvxq** [comment]
> @eugenov I'm usually used to testing new features in `Prerelease` versions. So there is currently no option to disable it. But I will provide such an option in the official version. I will also try my best to make `auto completion` perfect enough so that you won't think about closing it. 😄

let me show a reason of closing it, when I use it in gramma linux, and type 'll' command, and type enter, it always completes with kill ,  type space after ll can solve this problem , but I usually type enter after ll. when I choose by choosing none gramma, all color highlight disappeared. 
there is another question, I didn't find the option of remembering user name in a ssh session.

**Comment by printff** [comment]
Type ```cd``` then press enter, and will get ```chmod```.
Using right arrow instead Enter to select auto completion like mariadb client?

**Comment by yang-kai-shuo** [comment]
how to empty my history command？

**Comment by steinvenic** [comment]
> Type `cd` then press enter, and will get `chmod`. Using right arrow instead Enter to select auto completion like mariadb client?


**Auto Completion** also affects my normal full commands too,your suggest sounds good ~
![image](https://user-images.githubusercontent.com/18524326/176867955-0a9a294b-d7bc-41e4-936d-4720f67ea65f.png)

**Comment by ImMartinYoung** [comment]
2.5.0 Prerelease 3 is great, totally solving all related issues. This discussion may be closed.

**Comment by kingToolbox** [comment]
Sorry for replying so late, I just got back from a business trip.

@eugenov 
I still need time to perfect the format of the file, and it is expected that the customization of commands will be opened in subsequent versions. I look forward to volunteers adding commands such as `docker` commands, `switch/router` commands, etc.

@iciness 
@zdltvxq 
@printff 
@steinvenic
In the [WindTerm_2.5.0](https://github.com/kingToolbox/WindTerm/releases/tag/2.5.0), you can use the setting `Menu - Session - Preferences - Settings - Terminal - Auto Completion` to configure the `auto completion`:
![image](https://user-images.githubusercontent.com/51009775/186054204-8583f591-eee9-4c0d-9801-1945cd50d6ad.png)

@yang-kai-shuo
In the [WindTerm_2.5.0](https://github.com/kingToolbox/WindTerm/releases/tag/2.5.0), you can use the setting `Menu - Session - Preferences - Settings - Terminal - Command History` to clear the `command history`:
![image](https://user-images.githubusercontent.com/51009775/186054345-383d741b-7559-4721-a65b-92c285c5ea25.png)

@ImMartinYoung 
Thank you, because the autocomplete list item is no longer selected by default, users can type smoothly as before, and the matching algorithm has also been greatly improved to make the matching more accurate, so from the feedback, the current `auto completion` is not bad. :smile:

I will continue to improve it and still welcome any comments and suggestions. Thank you all.

**Comment by surenz** [comment]
Hi @kingToolbox, I haven't found any issue that relates to my case and I'm not sure it's a bug or feature but what happens for me is that if I'm authenticated already once, WindTerm tries to write the password on any 'password' prompt it gets (ssh-linux) which happens even when trying to change a password for user (i.e. 'passwd user'). This happens when OneKey autocompletion is enabled. If I disable it, the issue is gone. The WindTerm version is 2.7.0. if that matters. 
Is there a way to disable this first attempt to autocomplete authentication with enabled OneKey authentication autocomplete?
